File: citizensdk/darwin/Sources/CitizenSDK/CitizenSDKError.swift
import Foundation

/// Stable error vocabulary shared verbatim with `citizensdk_error_code_t`.
public enum CitizenSDKErrorCode: Int32, CaseIterable, Sendable {
    case ok = 0
    case invalidArgument = 1
    case invalidHandle = 2
    case invalidState = 3
    case unsupported = 4
    case unavailable = 5
    case notReady = 6
    case notFound = 7
    case conflict = 8
    case integrity = 9
    case authenticationCancelled = 10
    case authenticationRequired = 11
    case keyInvalidated = 12
    case permissionDenied = 13
    case storage = 14
    case network = 15
    case decode = 16
    case timeout = 17
    case busy = 18
    case queueFull = 19
    case internalFailure = 20
    case panic = 21
    case cancelled = 22

    internal static func checked(_ rawValue: Int32) -> CitizenSDKErrorCode {
        CitizenSDKErrorCode(rawValue: rawValue) ?? .integrity
    }
}

/// Public failures contain neither secrets nor Core request/result identities.
public struct CitizenSDKError: LocalizedError, Sendable, Equatable {
    public let code: CitizenSDKErrorCode
    public let message: String

    public init(_ code: CitizenSDKErrorCode, _ message: String) {
        self.code = code
        self.message = message
    }

    public var errorDescription: String? { message }
}

internal enum CitizenSDKChecks {
    static func requireOK(_ rawCode: Int32, _ fallback: String) throws {
        guard rawCode == CitizenSDKErrorCode.ok.rawValue else {
            throw CitizenSDKError(.checked(rawCode), CitizenSDKNative.lastError(fallback: fallback))
        }
    }

    static func require(_ condition: @autoclosure () -> Bool, _ message: String) throws {
        guard condition() else { throw CitizenSDKError(.invalidArgument, message) }
    }
}

File: citizensdk/android/native/src/main/kotlin/org/citizen/sdk/CitizenSdkError.kt
package org.citizen.sdk

/** Stable error vocabulary shared with `citizensdk_error_code_t`. */
enum class CitizenSdkErrorCode(val value: Int) {
    OK(0),
    INVALID_ARGUMENT(1),
    INVALID_HANDLE(2),
    INVALID_STATE(3),
    UNSUPPORTED(4),
    UNAVAILABLE(5),
    NOT_READY(6),
    NOT_FOUND(7),
    CONFLICT(8),
    INTEGRITY(9),
    AUTHENTICATION_CANCELLED(10),
    AUTHENTICATION_REQUIRED(11),
    KEY_INVALIDATED(12),
    PERMISSION_DENIED(13),
    STORAGE(14),
    NETWORK(15),
    DECODE(16),
    TIMEOUT(17),
    BUSY(18),
    QUEUE_FULL(19),
    INTERNAL(20),
    PANIC(21),
    CANCELLED(22);

    companion object {
        @JvmStatic
        fun fromValue(value: Int): CitizenSdkErrorCode =
            entries.firstOrNull { it.value == value } ?: INTEGRITY
    }
}

/** Public failures never contain secrets or native/result identities. */
class CitizenSdkException(
    val code: CitizenSdkErrorCode,
    message: String,
    cause: Throwable? = null,
) : RuntimeException(message, cause)
